crates/cast/src/lib.rs

```diff
@@ -981,6 +994,107 @@ impl<P: Provider<AnyNetwork>> Cast<P> {
         Ok(res)
     }
 
+    fn extract_block_range(filter: &Filter) -> Result<(Option<u64>, Option<u64>)> {
+        let FilterBlockOption::Range { from_block, to_block } = &filter.block_option else {
+            return Ok((None, None));
+        };
+
+        let extract_number = |block: Option<BlockNumberOrTag>| {
+            block.and_then(|b| match b {
+                BlockNumberOrTag::Number(n) => Some(n),
+                _ => None,
+            })
+        };
+
+        Ok((extract_number(*from_block), extract_number(*to_block)))
+    }
+
+    /// Retrieves logs with automatic chunking fallback.
+    ///
+    /// First tries to fetch logs for the entire range. If that fails,
+    /// falls back to concurrent chunked requests with rate limiting.
+    async fn get_logs_chunked(&self, filter: &Filter, chunk_size: u64) -> Result<Vec<Log>>
+    where
+        P: Clone + Unpin,
+    {
+        // Try the full range first
+        if let Ok(logs) = self.provider.get_logs(filter).await {
+            return Ok(logs);
+        }
```

Comment:
~~we should log the error here and we also need to consider that this might have failed for other reasons to provider better error responses, e.g. rn we might fail the request and then return an error if we can exract the range in get_logs_chunked_concurrent~~